### PR TITLE
Call calculate_tx_resources.

### PR DESCRIPTION
**Pull request by noaov1**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/395)

### CODE REVIEW

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 1 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/395)* status: 5 of 7 files reviewed, 11 unresolved discussions (waiting on @alonh5 and @noaov1)

___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 73 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSq1oy3CrEV9iFnyMiK:-NSq1oy3CrEV9iFnyMiL:b-va3s6n) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_utils.rs#L73)):*
> ```XML
> 
>     // TODO(Noa, 30/04/23): Consider adding builtin suffix in the VM or remove the suffix from the
>     // flow.
> ```

How difficult will it be to remove it from our flow? I imagine it appears once.



_Code quote:_
```XML
    // TODO(Noa, 30/04/23): Consider adding builtin suffix in the VM or remove the suffix from the
    // flow.
```
___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 83 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSq2NSsBmcSwUIb2CY_:-NSq2NSsBmcSwUIb2CYa:undefined) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_utils.rs#L83)):*
> ```XML
>     // Add additional Cairo resources needed for the OS to run the transaction.
>     let all_cairo_usage =
>         &cairo_usage + &get_additional_os_resources(resources_manager.syscall_counter, tx_type)?;
> ```





_Suggestion:_
```XML
    let total_cairo_usage =
        &cairo_usage + &get_additional_os_resources(resources_manager.syscall_counter, tx_type)?;
    let total_cairo_usage = all_cairo_usage.filter_unused_builtins();
```
___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 84 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSq2nHb1FNNbwMwuS2y:-NSq2nHb1FNNbwMwuS2z:b-rudhgy) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_utils.rs#L84)):*
> ```XML
>     let all_cairo_usage =
>         &cairo_usage + &get_additional_os_resources(resources_manager.syscall_counter, tx_type)?;
>     let all_cairo_usage_filtered = all_cairo_usage.filter_unused_builtins();
> ```

Perhaps fileterring unused should be done last?



_Code quote:_
```XML
let all_cairo_usage_filtered = all_cairo_usage.filter_unused_builtins();
```

**Review by noaov1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/395)* status: 5 of 7 files reviewed, 8 unresolved discussions (waiting on @alonh5 and @elintul)

___
*[`crates/blockifier/src/fee/fee_utils.rs` line 23 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSpqSod4IfVoqEwyqxC:-NSt637_CI1GCI_nPJ1z:b-4hvi0m) ([raw file](https://github.com/starkware-libs/blockifier/blob/a5dd7f77700b8a0cba49d69c34072dd4fe44fbb4/crates/blockifier/src/fee/fee_utils.rs#L23)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Code movement: either in a separate PR, or indicating that it was moved as-is (and to where).
In this case it was moved and changed, so I think moving could have been done in a separate PR for easier check.
</blockquote></details>

Got it.
___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 281 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSpqxnLBX_YIgJc7Yfo:-NSt6iueDZpXtUdespMx:bv5ip6a) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/account_transaction.rs#L281)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Can't we do `let tx_type = match self {...};`?
</blockquote></details>

Yes we can. We already have a `match self` so I thought the `tx_type` can be set inside it. Do you think that doing it separately is better?
___
*[`crates/blockifier/src/transaction/transaction_execution.rs` line 60 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSpsET57j67Ie5MAi6O:-NSt7M904IlRSKBOhJQj:batsk3j) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_execution.rs#L60)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Last PR of fees feature: remove these E2E.
</blockquote></details>

Got it. Thanks
___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 48 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSprAgD7gOBk5nzTaID:-NSt7V3ME6HqtXaRd18b:b-xumw1t) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_utils.rs#L48)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Why not explicitly pass `validate_call_info, execute_call_info`? It's the only two that are passed.
</blockquote></details>

In `L1HandlerTransaction` we only pass the `execute_call_info` but I we can pass `Option<CallInfo>`.  WDYT?
___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 73 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSq1oy3CrEV9iFnyMiK:-NStGpPS0Sug3v9L3Xru:bmhiway) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_utils.rs#L73)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

How difficult will it be to remove it from our flow? I imagine it appears once.
</blockquote></details>

The question is whether we want to remove it. The `cairo_resource_fee_weights` can contain things other than builtins: e.g. "n_steps"), so adding the suffix gives more information.
It is not difficult: we only need to remove it from `cairo_resource_fee_weights` that we get from python.
___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 84 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSq2nHb1FNNbwMwuS2y:-NStIpi665dIrlC3AwcH:b-hzkyoz) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_utils.rs#L84)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Perhaps fileterring unused should be done last?
</blockquote></details>

But then we'll copy it to `tx_resources` only to remove it later, no?
___
*[`crates/blockifier/src/transaction/transactions_test.rs` line 321 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSps0c0BWyWgksqCcd5:-NStJANS8hc4E6IrZ5uZ:bjcdufj) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transactions_test.rs#L321)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

These also exist as constants in `cairo-rs`, right?
</blockquote></details>

Correct (without the suffix). We can write: `HASH_BUILTIN_NAME.to_string() + BUILTIN_NAME_SUFFIX` but it is longer. WDYT?

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/395)* status: all files reviewed, 9 unresolved discussions (waiting on @alonh5 and @noaov1)

___
*[`crates/blockifier/src/abi/constants.rs` line 30 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NStTtKT22IzPBy7Gb6M:-NStTtKT22IzPBy7Gb6N:bahod39) ([raw file](https://github.com/starkware-libs/blockifier/blob/f1dde0dcf3b11b934f6e35604cf29aa36860cf10/crates/blockifier/src/abi/constants.rs#L30)):*
> ```XML
> // Transaction resource names.
> pub const GAS_USAGE: &str = "l1_gas_usage";
> pub const N_STEPS_RESOURCE: &str = "n_steps";
> ```

Do you think it's needed?
Given that the other is "gas usage" and not "gas usage resource", maybe we can stay with "n steps".



_Code quote:_
```XML
_RESOURCE
```
___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 281 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSpqxnLBX_YIgJc7Yfo:-NStU2aM-AbweYdw5XiO:bd32x3p) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/account_transaction.rs#L281)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Yes we can. We already have a `match self` so I thought the `tx_type` can be set inside it. Do you think that doing it separately is better?
</blockquote></details>

This whole method is pretty convoluted by now and will be refactored soon.
But I think it's better than `let X = None;` then setting `X` later.
___
*[`crates/blockifier/src/transaction/transaction_execution.rs` line 42 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NStUZHI6y2zEvR2mxz2:-NStUZHI6y2zEvR2mxz3:bknns6o) ([raw file](https://github.com/starkware-libs/blockifier/blob/f1dde0dcf3b11b934f6e35604cf29aa36860cf10/crates/blockifier/src/transaction/transaction_execution.rs#L42)):*
> ```XML
>             execution_resources,
>             execute_call_info.as_ref(),
>             None,
> ```

Give it a name (or an inline comment).



_Code quote:_
```XML
None,
```
___
*[`crates/blockifier/src/transaction/transaction_execution.rs` line 46 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NStUhKR4oolxg3lPQxC:-NStUhKR4oolxg3lPQxD:b-57mizy) ([raw file](https://github.com/starkware-libs/blockifier/blob/f1dde0dcf3b11b934f6e35604cf29aa36860cf10/crates/blockifier/src/transaction/transaction_execution.rs#L46)):*
> ```XML
>             state,
>             // The calldata includes the "from" field, which is not a part of the payload. We thus
>             // subtract 1.
> ```

Give this field a name too.



_Suggestion:_
```XML
            // The calldata includes the "from" field, which is not a part of the payload.
            Some(self.calldata.0.len() - 1),
```
___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 48 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSprAgD7gOBk5nzTaID:-NStUpXy8BUoKS3v4wdj:b-eq0prt) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_utils.rs#L48)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

In `L1HandlerTransaction` we only pass the `execute_call_info` but I we can pass `Option<CallInfo>`.  WDYT?
</blockquote></details>

I think it's better now, since most of the usages pass the two of them.
___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 73 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSq1oy3CrEV9iFnyMiK:-NStV8F23iAKikbB6BbH:bi5sb8n) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_utils.rs#L73)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

The question is whether we want to remove it. The `cairo_resource_fee_weights` can contain things other than builtins: e.g. "n_steps"), so adding the suffix gives more information.
It is not difficult: we only need to remove it from `cairo_resource_fee_weights` that we get from python.
</blockquote></details>

What is LC doing? Do they have those weight names without the suffix?
Maybe it's better to peel the suffix in `native-blockifier` code, and work without suffixes in `blockifier`.
___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 84 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSq2nHb1FNNbwMwuS2y:-NStVpEJEyHDBZ_vwdGw:b-uvelnh) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_utils.rs#L84)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

But then we'll copy it to `tx_resources` only to remove it later, no?
</blockquote></details>

Nevermind, I thought you were adding more builtin usage after filtering.
___
*[`crates/blockifier/src/transaction/transactions_test.rs` line 321 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSps0c0BWyWgksqCcd5:-NStULwU5AJ9isLAGng9:b9xxcpj) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transactions_test.rs#L321)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Correct (without the suffix). We can write: `HASH_BUILTIN_NAME.to_string() + BUILTIN_NAME_SUFFIX` but it is longer. WDYT?
</blockquote></details>

Answered above.
We can also do it in a separate PR.

**Review by noaov1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/395)* status: all files reviewed, 5 unresolved discussions (waiting on @alonh5 and @elintul)

___
*[`crates/blockifier/src/abi/constants.rs` line 30 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NStTtKT22IzPBy7Gb6M:-NStiVSiEH-K0QBb0NZM:bucaoev) ([raw file](https://github.com/starkware-libs/blockifier/blob/f1dde0dcf3b11b934f6e35604cf29aa36860cf10/crates/blockifier/src/abi/constants.rs#L30)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Do you think it's needed?
Given that the other is "gas usage" and not "gas usage resource", maybe we can stay with "n steps".
</blockquote></details>

It adds the information that we refer to the number of steps as a resource but maybe it's redundant. 
___
*[`crates/blockifier/src/transaction/transaction_execution.rs` line 42 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NStUZHI6y2zEvR2mxz2:-NStkK5d1lwHRq-qbZgM:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/f1dde0dcf3b11b934f6e35604cf29aa36860cf10/crates/blockifier/src/transaction/transaction_execution.rs#L42)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Give it a name (or an inline comment).
</blockquote></details>

Done.
___
*[`crates/blockifier/src/transaction/transaction_execution.rs` line 46 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NStUhKR4oolxg3lPQxC:-NStkKHfE-RuhQ_3IAY2:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/f1dde0dcf3b11b934f6e35604cf29aa36860cf10/crates/blockifier/src/transaction/transaction_execution.rs#L46)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Give this field a name too.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 42 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSprbWF2hmgiPekaT5W:-NStkLgG33jLT0iH79hx:b-m14d3z) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_utils.rs#L42)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

I don't think I fully understand this documentation. "As most recent" but then "as if it's the first"?
Can you rephrase it in a simpler way, or consider shortening?
</blockquote></details>

I deleted the last sentence (I asked Yoni and its is unnecessary and confusing). WDYT?

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/395)* status: all files reviewed, 5 unresolved discussions (waiting on @alonh5 and @noaov1)

___
*[`crates/blockifier/src/abi/constants.rs` line 30 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NStTtKT22IzPBy7Gb6M:-NStv6vDBWPvFcwTwYUC:brybs63) ([raw file](https://github.com/starkware-libs/blockifier/blob/f1dde0dcf3b11b934f6e35604cf29aa36860cf10/crates/blockifier/src/abi/constants.rs#L30)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

It adds the information that we refer to the number of steps as a resource but maybe it's redundant. 
</blockquote></details>

Consider adding this suffix to gas usage too.
___
*[`crates/blockifier/src/transaction/transaction_execution.rs` line 46 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NStUhKR4oolxg3lPQxC:-NStuUFyCghAPxjX5OND:bci97a2) ([raw file](https://github.com/starkware-libs/blockifier/blob/f1dde0dcf3b11b934f6e35604cf29aa36860cf10/crates/blockifier/src/transaction/transaction_execution.rs#L46)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Move the comment next to the new variable (without is last part)?
___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 42 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSprbWF2hmgiPekaT5W:-NStv18S44jSqaCGn-BC:bupwgau) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_utils.rs#L42)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

I deleted the last sentence (I asked Yoni and its is unnecessary and confusing). WDYT?
</blockquote></details>

Suggested small fixes.
___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 42 at r4](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NStugDk35KJzQPcdCsP:-NStugDk35KJzQPcdCsQ:undefined) ([raw file](https://github.com/starkware-libs/blockifier/blob/4dd5f5a55057d89135b3bdf94c192fba89fb90ad/crates/blockifier/src/transaction/transaction_utils.rs#L42)):*
> ```XML
> }
> 
> /// Returns the total resources needed to include the most recent transaction in a StarkNet batch
> ```





_Suggestion:_
```XML
/// Calculates the total resources needed to include the transaction in a StarkNet block as most-recent
/// (recent w.r.t. application on the given state).
/// I.e., L1 gas usage and Cairo VM execution resources.
```

**Review by noaov1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/395)* status: 5 of 7 files reviewed, 2 unresolved discussions (waiting on @alonh5, @elintul, and @noaov1)

___
*[`crates/blockifier/src/transaction/transaction_execution.rs` line 46 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NStUhKR4oolxg3lPQxC:-NStwVVU3lQTt9cCu9cW:b-l0b1qo) ([raw file](https://github.com/starkware-libs/blockifier/blob/f1dde0dcf3b11b934f6e35604cf29aa36860cf10/crates/blockifier/src/transaction/transaction_execution.rs#L46)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Move the comment next to the new variable (without is last part)?
</blockquote></details>

Oh missed that. Done.
___
*[`crates/blockifier/src/transaction/transaction_utils.rs` line 73 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/395#-NSq1oy3CrEV9iFnyMiK:-NStyaNZ32IxH8LagpJ5:b1lg3sh) ([raw file](https://github.com/starkware-libs/blockifier/blob/d455296ad0e74ff7384aabd09cd967e04afa8ff9/crates/blockifier/src/transaction/transaction_utils.rs#L73)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

What is LC doing? Do they have those weight names without the suffix?
Maybe it's better to peel the suffix in `native-blockifier` code, and work without suffixes in `blockifier`.
</blockquote></details>

Will do in another PR?
